### PR TITLE
Develop stream 2024-06-26

### LINKED ISSUE #570: Failing tests with libstdc++ assertions due to unsigned overflow
**Describe the bug**
Hi, while testing with hardened libstdc++ (compiled with `-D_GLIBCXX_ASSERTIONS`), many tests rocPRIM failed with traceback like:

```
libstdc++.so.6!std::__glibcxx_assert_fail(char const*, int, char const*, char const*) (Unknown Source:0)
std::uniform_int_distribution<unsigned int>::param_type::param_type() (\usr\lib\gcc\x86_64-pc-linux-gnu\13\include\g++-v13\bits\uniform_int_dist.h:108)
std::uniform_int_distribution<unsigned int>::uniform_int_distribution() (\usr\lib\gcc\x86_64-pc-linux-gnu\13\include\g++-v13\bits\uniform_int_dist.h:145)
test_utils::generate_random_data_n<__gnu_cxx::__normal_iterator<unsigned char*, std::vector<unsigned char, std::allocator<unsigned char> > >, int, int, std::linear_congruential_engine<unsigned long, 16807ul, 0ul, 2147483647ul> >() (\var\tmp\portage\sci-libs\rocPRIM-6.1.1\work\rocPRIM-rocm-6.1.1\test\rocprim\test_utils_data_generation.hpp:262)
test_utils::get_random_data<unsigned char, int, int>(seed_type seed_value) (\var\tmp\portage\sci-libs\rocPRIM-6.1.1\work\rocPRIM-rocm-6.1.1\test\rocprim\test_utils_data_generation.hpp:354)
RocprimBlockLoadStoreClassTestsFirstPart_LoadStoreClass_Test<class_params<unsigned char, (rocprim::block_load_method)2, (rocprim::block_store_method)2, 512u, 1u> >::TestBody(class RocprimBlockLoadStoreClassTestsFirstPart_LoadStoreClass_Test<class_params<unsigned char, (rocprim::block_load_method)2, (rocprim::block_store_method)2, 512u, 1u> > * this) (\var\tmp\portage\sci-libs\rocPRIM-6.1.1\work\rocPRIM-rocm-6.1.1\test\rocprim\test_block_load_store.hpp:53)
```

where starting point is `test_utils::get_random_data<Type>(size, -100, 100, seed_value);`
and ending point is `__glibcxx_assert(_M_a <= _M_b)`, where _M_a = 4294967196 and _M_b = 100.

I. e. tests try to build `uniform_int_distribution<unsigned int>` between -100 and 100, which results in implicit conversion, which results in [undefined behavior](https://stackoverflow.com/questions/63582797/uniform-int-distribution-with-zero-range-goes-to-infinite-loop).

Affected tests are:

- RocprimBlockLoadStoreClassTestsFirstPart/23.LoadStoreClass
- RocprimBlockSortMergeTestsIntegral/8.SortKeys
- RocprimBlockShuffleTestsIntegral/8.BlockOffset
- RocprimBlockSortBitonicTestsIntegral/8.SortKeys
- DeviceBatchMemcpyTests/11.SizeAndTypeVariation
- RocprimDeviceSortTests/0.SortKey
- RocprimDeviceRunLengthEncode/3.Encode
- RocprimDeviceUniqueByKeyTests/2.UniqueByKey

Some of them use unsigned short, some unsigned long long and so on. Replacing negative starting point with non-negative helps, but maybe you can provide a better solution?

**Environment**
- rocPRIM 6.1.1

### PR DESCRIPTION
**Pull request by Snektron**
@Naraenda is taking this over since @Snektron is on holiday.

This PR adds the following merge commits. **Emphasis** on important changes.

* 46faf89b **test: make tests pass with hardened libstc++** (resolves https://github.com/ROCm/rocPRIM/issues/570)
* 39c9d9bc ci(gitlab-ci.yml): re-enable most of the debug builds
* c6ea6aab test: improved test of device histogram_even for half/bfloat16
* a40cbdb9 fix: fix warnings when building rocprim from rocthrust when using hipcc
* 10c6175e **feat: add `partial_sort` and `partial_sort_copy` algorithms**
* 4a268483 **feat: add deterministic versions of `rocprim::inclusive_scan`, `rocprim::exclusive_scan`, `rocprim::inclusive_scan_by_key`, `rocprim::exclusive_scan_by_key`, and `rocprim::reduce_by_key`**
* 9c7c2970 style: format attributes on a separate line
* fc2de523 perf: remove oracle in nth element
* a4ae29cb ci(gitlab-ci.yml): compress autotuning artifacts before uploading them
* fab7fd63 refactor: improve code-reuse of `rocprim::detail::merge_path`
* f586d3ed test: add extra tests to nth-element
* d09bae6e **feat: add nth-element algorithm**
* 43c7ce71 perf: add auto tuning to reduce by key
* d63cd11a **feat: add large index support to device segmented reduce**
* 110f2d0f test: fix sizes generated by benchmark utility for random segments
* c523e13f ci: build everything using cmake-minimum and cmake-latest workflows
* d4aec9d4 test: fix benchmark minmax underoverflow
* 941984e4 test: disable naive batch memcpy in benchmarks
* 94e17b48 docs: document thread level methods
* f90e4e7e refactor: only autotune sleep-based lookback scan for gfx908
* 7b6b04bd docs: add documentation for autotuning
* e2aec7e6 test: improve test logging
* 1617f92a **refactor: deprecate raw_storage and replace it by uninitialized_array due to possible undefined behavior with raw_storage**
* f7209def **refactor: deprecate use of rocprim::thread_load and rocprim::thread_store**
* 917b1510 fix: fix block histogram test failures
* 441a934f docs: document thread types
* 2455a40e ci(gitlab-ci.yml): disable debug builds due to long compilation times
* 9fc5d182 fix: remove memory fence workaround on gfx10 and gfx11
* c07e212a fix: fix unroll warnings in debug min size releases
* ad746b46 **perf: add tuning to device partition**
* 9148c771 test: add benchmark flag to specify seed
* 9e849dbd docs: add contribution guidelines
* 71ba9cef **fix: prevent rocprim::adjacent_difference benchmarks from using more than 8 gb vram**
* d3b34418 test: fix windows test assertions
* c5a9d07c fix: disable force inline on windows
* 58c96d95 refactor: remove device device warpsize usage in host function in radix sort benchmark generator
* fc2ac733 **feat: add an overload of rocprim::match_any which accepts the maximum number of bits as a run-time parameter**
* 61ca0d13 **test: change default size in adjacent difference benchmark**
* 246fe37d fix: various fixes for batch memcpy
* de58f352 **perf: add auto tuning to transform**

### CODE REVIEW

**Comment by Naraenda** [comment]
Rebased and added a fix that resolves https://github.com/ROCm/rocPRIM/issues/570

**Comment by Naraenda** [comment]
Fixed doxygen errors in https://github.com/ROCm/rocPRIM/pull/575/commits/8d5b5e89021b08cb51d783cf0021294f545fac4c

**Comment by Naraenda** [comment]
4651b07 Noticed CI failures after targeting a later commit in develop. `rocm-cmake` was recently-ish changed and now requires building. This commit resolves that when the `rocm-cmake` is freshly downloaded through `FetchContent`.

6edfe13 Should fix one of the test failures on MI200 with very large inputs.

Also rebased it, again.

CC @stanleytsang-amd

**Comment by stanleytsang-amd** [comment]
> [4651b07](https://github.com/ROCm/rocPRIM/commit/4651b07a6af1642f0d0bc9d7166b4298d8eb4ade) Noticed CI failures after targeting a later commit in develop. `rocm-cmake` was recently-ish changed and now requires building. This commit resolves that when the `rocm-cmake` is freshly downloaded through `FetchContent`.
> 
> [6edfe13](https://github.com/ROCm/rocPRIM/commit/6edfe13e6c54fc564bf786200740410d962f36f4) Should fix one of the test failures on MI200 with very large inputs.
> 
> Also rebased it, again.
> 
> CC @stanleytsang-amd

Thanks Nara, I will rerun CI right now.

**Comment by Naraenda** [comment]
I managed to test this on a MI200. It turns out, for reasons I'm not fully aware of, that running `hipStreamSynchronize` after the cooperative kernel launch is required. Moving that statement out of the debug synchronous branch seems to resolve the hanging on MI200 (4314c76). I'm checking the implementation in ROCm/CLR, but nothing out of the ordinary seems to jump out.

CC @stanleytsang-amd

**Comment by stanleytsang-amd** [comment]
> I managed to test this on a MI200. It turns out, for reasons I'm not fully aware of, that running `hipStreamSynchronize` after the cooperative kernel launch is required. Moving that statement out of the debug synchronous branch seems to resolve the hanging on MI200 ([4314c76](https://github.com/ROCm/rocPRIM/commit/4314c76ea790369a5b7819515aff5f5274c900aa)). I'm checking the implementation in ROCm/CLR, but nothing out of the ordinary seems to jump out.
> 
> CC @stanleytsang-amd

Re-running CI now.  It does seem strange that an explicit hipStreamSynchronize is required here since this function only deals with one stream and all kernels are being executed on that stream and I don't see any data race conditions here.  Maybe we can make a small reproducer and file a bug report with the HIP team?

**Comment by stanleytsang-amd** [comment]
> I managed to test this on a MI200. It turns out, for reasons I'm not fully aware of, that running `hipStreamSynchronize` after the cooperative kernel launch is required. Moving that statement out of the debug synchronous branch seems to resolve the hanging on MI200 ([4314c76](https://github.com/ROCm/rocPRIM/commit/4314c76ea790369a5b7819515aff5f5274c900aa)). I'm checking the implementation in ROCm/CLR, but nothing out of the ordinary seems to jump out.
> 
> CC @stanleytsang-amd

It seems as though the hang still persists on MI200 on CI, even with the hipStreamSychronize().

**Comment by Naraenda** [comment]
Rebased & resolved conflicts

**Review by stanleytsang-amd — CHANGES_REQUESTED** [review]
Can you update the packaging version at https://github.com/ROCm/rocPRIM/pull/575/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL133 to 3.3.0 as well? Thanks.

**Comment by Naraenda** [comment]
> Can you update the packaging version at https://github.com/ROCm/rocPRIM/pull/575/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL133 to 3.3.0 as well? Thanks.

Done in 8d06855.

**Comment by stanleytsang-amd** [comment]
> > Can you update the packaging version at https://github.com/ROCm/rocPRIM/pull/575/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aL133 to 3.3.0 as well? Thanks.
> 
> Done in [8d06855](https://github.com/ROCm/rocPRIM/commit/8d068554f5fe7cac67dda25d1f21723be7aadfd2).

Thanks.

**Review by stanleytsang-amd — APPROVED** [review]

File: benchmark/benchmark_config_dispatch.cpp [hunk 1]
```diff
@@ -91,20 +91,19 @@ int main(int argc, char** argv)
...
     std::vector<benchmark::internal::Benchmark*> benchmarks{
         CREATE_BENCHMARK(default_stream, stream_kind::default_stream),
         CREATE_BENCHMARK(per_thread_stream, stream_kind::per_thread_stream),
         CREATE_BENCHMARK(explicit_stream, stream_kind::explicit_stream),
-        CREATE_BENCHMARK(async_stream, stream_kind::async_stream)
-    };
-
+        CREATE_BENCHMARK(async_stream, stream_kind::async_stream),
+        benchmark::RegisterBenchmark(
+            bench_naming::format_name("{lvl:na,algo:empty_kernel,cfg:default_config}"),
```
Comment (stanleytsang-amd) [inline benchmark/benchmark_config_dispatch.cpp]: This does not compile:

/tmp/grpc/rocPRIM/benchmark/benchmark_config_dispatch.cpp:104:9: error: no matching function for call to 'RegisterBenchmark'
  104 |         benchmark::RegisterBenchmark(
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/benchmark/benchmark.h:1182:29: note: candidate function not viable: no known conversion from 'std::string' (aka 'basic_string<char>') to 'const char *' for 1st argument
 1182 | inline internal::Benchmark* RegisterBenchmark(const char* name,
      |                             ^                 ~~~~~~~~~~~~~~~~
/usr/local/include/benchmark/benchmark.h:1190:22: note: candidate function template not viable: no known conversion from 'std::string' (aka 'basic_string<char>') to 'const char *' for 1st argument
 1190 | internal::Benchmark* RegisterBenchmark(const char* name, Lambda&& fn) {
      |                      ^                 ~~~~~~~~~~~~~~~~
/usr/local/include/benchmark/benchmark.h:1201:22: note: candidate function template not viable: no known conversion from 'std::string' (aka 'basic_string<char>') to 'const char *' for 1st argument
 1201 | internal::Benchmark* RegisterBenchmark(const char* name, Lambda&& fn,

Adding `.c_str()` fixes the compile error.

**Review by stanleytsang-amd — CHANGES_REQUESTED** [review]

File: benchmark/benchmark_config_dispatch.cpp (hunk 1 above)
Comment (Beanavil) [reply benchmark/benchmark_config_dispatch.cpp]: Should be fixed now!

**Review by stanleytsang-amd — APPROVED** [review]

**Pull request merged**